src/peft/tuners/reft/layer.py

```diff
@@ -0,0 +1,262 @@
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import warnings
+from typing import Any, List, Optional, Set, Tuple
+
+import torch
+import torch.nn as nn
+
+from peft.tuners.lycoris_utils import LycorisLayer, check_adapters_to_merge
+
+class LowRankRotateLayer(torch.nn.Module):
+    """A linear transformation with orthogonal initialization."""
+
+    def __init__(self, n, m):
+        super().__init__()
+        # n > m
+        self.weight = torch.nn.Parameter(torch.empty(n, m), requires_grad=True)
+        torch.nn.init.orthogonal_(self.weight)
+
+    def forward(self, x):
+        return torch.matmul(x.to(self.weight.dtype), self.weight)
+
+
+
+class ReftLayer(nn.Module, LycorisLayer):
+    # All names of layers that may contain adapter weights
+    adapter_layer_names = ("reft_A", "reft_R")
+    # other_param_names is defined on parent class
+
+    def __init__(self, base_layer: nn.Module):
+        super().__init__()
+        LycorisLayer.__init__(self, base_layer)
+
+        # LoHa info
+        self.reft_A = nn.ModuleDict({})
+        self.reft_R = nn.ModuleDict({})
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, nn.Conv2d):
+            in_features, out_features = base_layer.in_channels, base_layer.out_channels
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+
+    @property
+    def _available_adapters(self) -> Set[str]:
+        return {*self.reft_A, *self.reft_R}
+
+    def create_adapter_parameters(self, adapter_name: str, r: int, shape: Tuple[int, ...]):
+        rotate_layer = LowRankRotateLayer(self.out_features, r)
+        self.reft_R[adapter_name] = torch.nn.utils.parametrizations.orthogonal(rotate_layer, orthogonal_map='cayley')
```

Comment:
Is this redundant with `torch.nn.init.orthogonal_` above?

Reply:
`torch.nn.init.orthogonal_` ensures that the tensor is orthogonal at once, and `torch.nn.utils.parametrizations.orthogonal` ensures that the tensor is orthogonal during the training. 
I checked the impelementation of `torch.nn.utils.register_parametrization`. It calls the a parameterization function (`orthogonal_`) in the initilization process. Thus, `torch.nn.init.orthogonal_` is redundant.

Reply:
So IIUC, we can remove one of the invocations?

Reply:
Yes, we can remove one of them.